ls1intum/Artemis
# `Assessment`: Calculate exam grade on the server

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Currently the calculation logic for exam grades and points is on the client side. The aim of this PR is correcting the responsibility assignments, unifying the execution of the calculations by having it only on the server and not duplicating it on the client and simplifying the client code. However it is also important to ensure performance is at a feasible level and the server is not burdened unnecessarily.

### Description

The calculations for max, bonus and achieved points per exercise and their summations are moved to server side. The extra request made to `/courses/{courseId}/grading-scale/grade-steps/match-grade-step` is removed from exam score related components so all the results are returned to the client in a single request as opposed to 2 requests. The current observed functionality is preserved except that the student exam detail view now shows the max points of the exercises included as bonus in the `Bonus Points` column instead of the `Points` column, which seems more intuitive.

Thanks to @b-fein, the calculations are also tested with [PIT Mutation Testing](https://pitest.org/)

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with normal and bonus exercises

1. Log in to Artemis
2. Navigate to Course Administration
3. As Student, go to the exam and make a valid submission.
4. As Instructor, grade the submission.
5. As Instructor, go to student exam detail view and see that the max points of the exercises included as bonus are shown in the `Bonus Points` column instead of the `Points` column (See screenshots below).
6. As Student, review the exam grade and confirm exam related functionalities working as before except the bonus points mentioned in the previous item.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| ExamService.java | 80% | 93% |
| StudentExamResource.java | 78% | 92% |
| ExamScoresDTO.java | n/a | 100% |
| StudentExamWithGradeDTO.java | n/a | 100% |
| exam-management-resolve.service.ts | 88.37% | 100% |
| student-exam-detail-table-row.component.ts | 72.5% | 87.17% |
| student-exam-detail.component.ts | 61.79% | 86.4% |
| student-exam-summary.component.ts | 75% | 100% |
| student-exam.service.ts | 86.36% | 100% |
| exam-participation.component.ts | 63.95% | 77.43% |
| exam-participation.service.ts | 66.66% | 82.3% |
| exam-participation-summary.component.ts | 67.27% | 86.11% |
| exam-points-summary.component.ts | 90.69% | 100% |

### Screenshots



Old Student Exam Detail View
![StudentExamDetailComponent_develop_cropped_20220526](https://user-images.githubusercontent.com/18427209/170705758-085a884c-0927-4fd3-84d0-8bdbdb7394c2.png)

New Student Exam Detail View
![StudentExamDetailComponent_myExpectation_cropped_20220526](https://user-images.githubusercontent.com/18427209/170705802-9c11c694-02ef-471f-b25f-d75cbc6456c1.png)
